src/libraries/Common/src/Internal/Cryptography/PemKeyImportHelpers.cs

```diff
@@ -0,0 +1,157 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography;
+
+namespace Internal.Cryptography
+{
+    internal static class PemKeyImportHelpers
+    {
+        public delegate void ImportKeyAction(ReadOnlySpan<byte> source, out int bytesRead);
+        public delegate ImportKeyAction? FindImportActionFunc(ReadOnlySpan<char> label);
+        public delegate void ImportEncryptedKeyAction<TPass>(
+            ReadOnlySpan<TPass> password,
+            ReadOnlySpan<byte> source,
+            out int bytesRead);
+
+        public static void ImportEncryptedPem<TPass>(
+            ReadOnlySpan<char> input,
+            ReadOnlySpan<TPass> password,
+            ImportEncryptedKeyAction<TPass> importAction)
+        {
+            bool foundEncryptedPem = false;
+            PemFields foundFields = default;
+            ReadOnlySpan<char> foundSlice = default;
+
+            ReadOnlySpan<char> pem = input;
+            while (PemEncoding.TryFind(pem, out PemFields fields))
+            {
+                ReadOnlySpan<char> label = pem[fields.Label];
+
+                if (label.SequenceEqual("ENCRYPTED PRIVATE KEY"))
+                {
+                    if (foundEncryptedPem)
+                    {
+                        throw new ArgumentException(SR.Argument_PemImport_AmbiguousPem, nameof(input));
+                    }
+
+                    foundEncryptedPem = true;
+                    foundFields = fields;
+                    foundSlice = pem;
+                }
+
+                Index offset = fields.Location.End;
+                pem = pem[offset..];
+            }
+
+            if (!foundEncryptedPem)
+            {
+                throw new ArgumentException(SR.Argument_PemImport_NoPemFound, nameof(input));
+            }
+
+            ReadOnlySpan<char> base64Contents = foundSlice[foundFields.Base64Data];
+            int base64size = foundFields.DecodedDataLength;
+            byte[] decodeBuffer = CryptoPool.Rent(base64size);
+
+            try
+            {
+                if (!Convert.TryFromBase64Chars(base64Contents, decodeBuffer, out int bytesWritten))
+                {
+                    // Couldn't decode base64. We shouldn't get here since the
+                    // contents are pre-validated.
+                    Debug.Fail("Base64 decoding failed on already validated contents.");
+                    throw new ArgumentException();
+                }
+
+                Debug.Assert(bytesWritten == base64size);
+                Span<byte> decodedBase64 = decodeBuffer.AsSpan(0, base64size);
```

Comment:
In bizarro-world, where TryFromBase64Chars reports a different answer in a release build, let's use the answer it gave.

```suggestion
                Span<byte> decodedBase64 = decodeBuffer.AsSpan(0, bytesWritten);
```